### PR TITLE
Add SageMakerTrainer for model paralellism

### PR DESCRIPTION
**Pull request by sgugger**
# What does this PR do?

This PR adds a subclass of `Trainer` to use model parallelism in SageMaker. This new `Trainer` still retains all the functionality of the previous trainer (e.g. when not run in SageMaker it will work like the normal one) while automatically enabling model parallelism when an example script is launched via SageMaker with that option activated.

The easiest way to enable this in any example script is to replace the `Trainer` and `TrainingArguments` imports by:
```python
from transformers.sagemaker import SageMakerTrainingArguments as TrainingArguments, SageMakerTrainer as Trainer
```

Along the way, I had to refactor a few things in `Trainer` to make it easier to deal with stuff in the subclass (without having to rewrite the whole train method for instance), mainly the part that does the model wrapping. Also, there was a subtle bug coming from the fact SageMaker wrapper for the model for model parallelism changes the forward method of the model, so the `Trainer` will now store the arguments in the signature (in case that signature changes after wrapping the model).

### CODE REVIEW

File: src/transformers/sagemaker/trainer_sm.py [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+            else:
+                return DistributedSampler(self.train_dataset, num_replicas=smp.dp_size(), rank=smp.dp_rank())
+        else:
+            return super()._get_train_sampler()
+
+    def _get_eval_sampler(self, eval_dataset: Dataset) -> Optional[torch.utils.data.sampler.Sampler]:
+        if self.is_model_parallel_enabled:
+            return SequentialDistributedSampler(eval_dataset, num_replicas=smp.dp_size(), rank=smp.dp_rank())
+        else:
+            return super()._get_eval_sampler(eval_dataset)
```
Comment (sgugger) [inline src/transformers/sagemaker/trainer_sm.py:93]: Same behavior as regular `Trainer` but we need to `smp.dp_size()` and `smp.dp_rank()` for the data parallelism (some processes use the same data if they are on the model parallelism axis).

File: src/transformers/sagemaker/trainer_sm.py [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+            model.train()
+            inputs = self._prepare_inputs(inputs)
+            loss_mb = forward_backward(model, inputs, self.args.gradient_accumulation_steps)
+            return loss_mb.reduce_mean().detach().to(self.args.device)
+        else:
+            return super().training_step(model, inputs)
+
+    def _gather_and_numpify(self, tensors, name):
+        if self.is_model_parallel_enabled:
+            tensors = smp_gather(tensors)
```
Comment (sgugger) [inline src/transformers/sagemaker/trainer_sm.py:120]: Need a special gather because we don't gather across all processes, just the ones on the data parallelism axis.

File: src/transformers/sagemaker/trainer_sm.py [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+        elif isinstance(tensor, dict):
+            return type(tensor)({k: smp_gather(v) for k, v in tensor.items()})
+        elif not isinstance(tensor, torch.Tensor):
+            raise TypeError(
+                f"Can't gather the values of type {type(tensor)}, only of nested list/tuple/dicts of tensors."
+            )
+        all_tensors = smp.allgather(tensor, smp.CommGroup.DP_GROUP)
+        return torch.cat([t.cpu() for t in all_tensors], dim=0)
+
+    def nested_smp_concat(tensor):
```
Comment (sgugger) [inline src/transformers/sagemaker/trainer_sm.py:63]: This method takes outputs from the `forward_only` method and (recursively) concatenates them in one tensor.

File: src/transformers/sagemaker/trainer_sm.py [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+        if isinstance(tensor, (list, tuple)):
+            return type(tensor)(smp_gather(t) for t in tensor)
+        elif isinstance(tensor, dict):
+            return type(tensor)({k: smp_gather(v) for k, v in tensor.items()})
+        elif not isinstance(tensor, torch.Tensor):
+            raise TypeError(
+                f"Can't gather the values of type {type(tensor)}, only of nested list/tuple/dicts of tensors."
+            )
+        all_tensors = smp.allgather(tensor, smp.CommGroup.DP_GROUP)
+        return torch.cat([t.cpu() for t in all_tensors], dim=0)
```
Comment (sgugger) [inline src/transformers/sagemaker/trainer_sm.py:61]: Needed because model parallel ends up with tensors across multiple devices.

File: src/transformers/sagemaker/trainer_sm.py [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+
+    def nested_smp_concat(tensor):
+        if isinstance(tensor, (list, tuple)):
+            return type(tensor)(nested_smp_concat(t) for t in tensor)
+        elif isinstance(tensor, dict):
+            return type(tensor)({k: nested_smp_concat(v) for k, v in tensor.items()})
+        # It doesn't seem possible to check here if `tensor` is a StepOutput because StepOutput lives in `smp.step`
+        # which is also the name of the decorator so Python is confused.
+        return tensor.concat().detach().cpu()
```
Comment (sgugger) [inline src/transformers/sagemaker/trainer_sm.py:70]: `cpu` needed because model parallel ends up with tensors across multiple devices.

File: src/transformers/sagemaker/trainer_sm.py [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+        super().create_optimizer_and_scheduler(num_training_steps)
+        if self.is_model_parallel_enabled:
+            self.optimizer = smp.DistributedOptimizer(self.optimizer)
+
+    def training_step(self, model: nn.Module, inputs: Dict[str, Union[torch.Tensor, Any]]) -> torch.Tensor:
+        if self.is_model_parallel_enabled:
+            model.train()
+            inputs = self._prepare_inputs(inputs)
+            loss_mb = forward_backward(model, inputs, self.args.gradient_accumulation_steps)
+            return loss_mb.reduce_mean().detach().to(self.args.device)
```
Comment (sgugger) [inline src/transformers/sagemaker/trainer_sm.py:114]: The cast to device is needed because model parallel ends up with tensors across multiple devices and the training losses are accumulated on the device in the main Trainer (to avoid unnecessary `.item` on TPU).

File: src/transformers/sagemaker/training_args_sm.py [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+            self._n_gpu = 1
+
+        if device.type == "cuda":
+            torch.cuda.set_device(device)
+
+        return device
+
+    @property
+    def place_model_on_device(self):
+        return not (is_smdistributed_available() and self.mp_parameters != "")
```
Comment (sgugger) [inline src/transformers/sagemaker/training_args_sm.py:89]: Needed this new argument to now if I should do the `model.to(device)` inside Trainer (which should usually be the case, but not for model parallelism).

**Review by LysandreJik — APPROVED** [review]
LGTM! Great job @sgugger

File: src/transformers/sagemaker/training_args_sm.py [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+    return importlib.util.find_spec("smdistributed") is not None
+
+
+if is_smdistributed_available():
+    import smdistributed.modelparallel.torch as smp
+
+
+@dataclass
+class SageMakerTrainingArguments(TrainingArguments):
+    mp_parameters: str = field(
```
Comment (anirudh2290) [inline src/transformers/sagemaker/training_args_sm.py:38]: Where is this being set ?

**Review by anirudh2290 — COMMENTED** [review]
Thanks for the nice work @sgugger! I added some comments. Feel free to ignore if they don't make much sense.

File: src/transformers/sagemaker/trainer_sm.py (hunk 1 above)
Comment (mansimane) [reply src/transformers/sagemaker/trainer_sm.py:93]: When `is_model_parallel_enabled` and iterable datset is used, this will give error that iterable dataset does not have lenght because distributed sampler queries for it. We should also handle iterable dataset case separately here as done in main trainer class.

Comment (sgugger) [reply src/transformers/sagemaker/trainer_sm.py:93]: Iterable dataset in a distributed training is not really supported in the regular Trainer (in the sense that all the processes will see the same data) so I'm fine with an error being raised here, since this is a distributed setup.

File: src/transformers/sagemaker/trainer_sm.py (hunk 3 above)
Comment (sgugger) [reply src/transformers/sagemaker/trainer_sm.py:63]: The function is only used for outputs from our models, so list/tuple/dict of `StepOutput`. There might be a second level of nestedness (e.g. a tuple containing one tensor and a tuple of tensors) if the user set `output_hidden_states=True` or `output_attentions=True` but I doubt a user will get there.

File: src/transformers/sagemaker/trainer_sm.py [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+
+
+logger = logging.get_logger(__name__)
+
+
+if is_smdistributed_available():
+    import smdistributed.modelparallel.torch as smp
+
+    @smp.step()
+    def forward_backward(model, inputs, gradient_accumulation_steps=1):
```
[the remark this answers is not in the document]
Comment (sgugger) [reply src/transformers/sagemaker/trainer_sm.py:40]: Will address by erroring at init and remove that argument here.

File: src/transformers/sagemaker/training_args_sm.py (hunk 8 above)
Comment (sgugger) [reply src/transformers/sagemaker/training_args_sm.py:38]: This is an argument needed to be able to launch the script via SageMaker (if it's not accepted, the script will error out). We have a class that converts dataclasses like this in an argument parser, so this line here is what will make all the example scripts (or any script using `SageMakerTrainer`) accept `--mp_parameters` from the CLI.

Unless I was launching my test scripts the wrong way?

Comment (anirudh2290) [reply src/transformers/sagemaker/training_args_sm.py:38]: Aah got it. Makes sense.

File: src/transformers/sagemaker/trainer_sm.py (hunk 1 above)
Comment (sgugger) [reply src/transformers/sagemaker/trainer_sm.py:93]: (Of course when we roll out support for iterable datasets for distributed training, we will definitely add a special case here.)

**Review by anirudh2290 — APPROVED** [review]

File: src/transformers/sagemaker/trainer_sm.py (hunk 1 above)
Comment (mansimane) [reply src/transformers/sagemaker/trainer_sm.py:94]: @sgugger  For Iterable dataset, logic to make sure that all processes don’t see  the same data needs to be implemented in the dataset. The main trainer class still supports iterable dataset - https://github.com/huggingface/transformers/blob/1c8c2d9ab34b8c8d326db9e0608f8e54cfccb885/src/transformers/trainer.py#L446.  Perhaps we can  handle it similarly here and print a warning to take care of data duplication in dataset?
